internal/service/transaction_service.go

```diff
@@ -68,3 +71,74 @@ func (r *transactionServiceImpl) Deposit(ctx context.Context, owner string, amou
 
 	return r.transactionRepo.Create(ctx, transaction)
 }
+
+func (r *transactionServiceImpl) Withdraw(ctx context.Context, owner string, amount float32) error {
+	_, err := r.accountRepo.FindOne(ctx, owner)
+	if err != nil {
+		return err
+	}
+
+	err = r.debit(ctx, owner, owner, amount)
+	if err != nil {
+		return err
+	}
+
+	transaction := models.Transaction{
+		CreatedAt:  clockNow(),
+		IsConsumed: true,
+		Owner:      owner,
+		Sender:     owner,
+		Receiver:   owner,
+		Amount:     amount,
+	}
+	return r.transactionRepo.Create(ctx, transaction)
+}
+
+func (r *transactionServiceImpl) debit(ctx context.Context, owner string, receiver string, amount float32) error {
+	if amount >= 0 {
+		return ErrInvalidAmount
+	}
+
+	var oldest models.Transaction
+	transactions, err := r.transactionRepo.FindAll(ctx, owner)
+	if err != nil {
+		return err
+	}
+
+	if len(transactions) == 0 {
+		return ErrNoTransactions
+	}
+
+	var balance float64
+	for _, t := range transactions {
+		if t.Owner == owner && !t.IsConsumed {
+			balance += float64(t.Amount)
+			if (models.Transaction{} == oldest) || t.CreatedAt.Before(oldest.CreatedAt) {
+				oldest = t
+			}
+		}
+	}
+
+	if (balance + float64(amount)) < 0 {
+		return ErrInsufficentBalance
+	}
+
+	err = r.transactionRepo.MarkAsConsumed(ctx, oldest.TransactionId)
+	if err != nil {
+		return err
+	}
+
+	if (oldest.Amount + amount) != 0 {
```

Comment:
Theorically, the amount can consume more than one transaction.
Imagine this: 
- you 3 transaction of 200, 100 and 300
- you withdraw 400
- it should consume the 200, 100, and partially the 300